daemon/containerd/image_builder.go

```diff
@@ -440,20 +442,32 @@ func (i *ImageService) CreateImage(ctx context.Context, config []byte, parent st
 		layers = parentImageManifest.Layers
 	}
 
-	// get the info for the new layers
-	info, err := i.client.ContentStore().Info(ctx, layerDigest)
+	cs := i.client.ContentStore()
+
+	ra, err := cs.ReaderAt(ctx, ocispec.Descriptor{Digest: layerDigest})
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to read diff archive: %w", err)
 	}
+	defer ra.Close()
 
-	// append the new layer descriptor
-	layers = append(layers,
-		ocispec.Descriptor{
-			MediaType: containerdimages.MediaTypeDockerSchema2LayerGzip,
-			Digest:    layerDigest,
-			Size:      info.Size,
-		},
-	)
+	empty, err := archive.IsEmpty(content.NewReader(ra))
+	if err != nil {
+		return nil, fmt.Errorf("failed to check if archive is empty: %w", err)
+	}
+	if !empty {
+		info, err := cs.Info(ctx, layerDigest)
+		if err != nil {
+			return nil, err
+		}
+
+		layers = append(layers,
+			ocispec.Descriptor{
+				MediaType: containerdimages.MediaTypeDockerSchema2LayerGzip,
+				Digest:    layerDigest,
+				Size:      info.Size,
+			},
+		)
```

Comment:
Oh, Guess this change got lost somewhere along the way (not a blocker, but I think you modified it in an earlier iteration);

```suggestion
		layers = append(layers, ocispec.Descriptor{
			MediaType: containerdimages.MediaTypeDockerSchema2LayerGzip,
			Digest:    layerDigest,
			Size:      info.Size,
		})
```